### ISSUE TITLE
Embedding fonts

### ISSUE THREAD

**Issue by simonbunge** [gitbrent/PptxGenJS#176]
Any appetite for adding embedding fonts a feature? The use case is using custom-made (corporate) font faces, which the client obviously wouldn't have installed locally.

Some initial thinking around implementation...

**1. ...\ppt\slide\slideN.xml**

Currently, PptxGenJS generates tags like this:

`<a:latin charset="0" pitchFamily="34" typeface="Courier New"/>`

In a file generated by PowerPoint 2013 with an embedded font ("Custom1" as an example typeface name) this looks like:

`<a:latin charset="0" pitchFamily="2" panose="02000503020000020004" typeface="Custom1"/>`

The panose tag appears to be optional in the relevant XSD, which would mean no change required on that level.

**2. ...\ppt\presentation.xml**

This would need to be amendeded with the following structure.

```
<p:embeddedFontLst>
	<p:embeddedFont>
		<p:font charset="0" pitchFamily="2" panose="02000503020000020004" typeface="Custom1"/>
		<p:regular r:id="rId7"/>
		<p:bold r:id="rId8"/>
		<p:italic r:id="rId9"/>
		<p:boldItalic r:id="rId10"/>
	</p:embeddedFont>
</p:embeddedFontLst>
```

**3. ...\ppt\fonts\\**

Would need to add a folder with Embedded OpenType (EOT) .fntdata files in the archive. These binary structures could be manually be extracted from a corporate template and stored on the server to be pulled by the script through a URL. Interestingly, the above 'Custom 1' example is contained in font8.fntdata, so I don't quite understand how the referencing is done.

**Comment by gitbrent** [comment]
Hi @simonbunge ,

Thanks for the question/feature request.  I really like the way you've researched and presented the requirements.

I'm not averse to the idea of embedding fonts - when I worked at Dell we had to use "Museo Sans for Dell", so i'm familiar with the pains involved when sending PPT's to clients.

Since the EOT font files would have to be read/zipped, this solution would only apply to Node.js -
 unless we could utilize base64 via `data` like we do for images - the rest of the changes seem straight forward enough.

I'll mark this Issue as help wanted to see if anyone would like to contribute a patch tor provide this functionality.  I may be able to look into implementing it, but not in the immediate future.

**Comment by simonbunge** [comment]
Thanks, @gitbrent,

really impressed with your responsiveness and the project in general!

I think this can work with a purely client-side solution (other than potentially creating a 'dumb' server side container for the font binaries). I'll play with the code and see where it goes...

**Comment by NeomMob** [comment]
Why cant' we embbed a font using web client with a local path like the method used for images?
+1 for the need of this feature and +1 for your awesome support!

**Comment by simonbunge** [comment]
Done a bit of research. Turns out things are not as easy as they seemed. In fact - spoiler alert - I don't think that feature can reasonably be implemented without access to MS's code base. Here's the story..

So reading a base64 stream from a .ttf file, converting it to eot (see https://github.com/fontello/ttf2eot) and including it in the .pptx archive would be easy enough. Add a bit of more or less trivial 'plumbing' in the main XML and... bingo!

The problem is that PowerPoint, unlike other Office applications, doesn't appear to support the embedding of full ttf/eot fonts as such, see https://docs.microsoft.com/en-us/openspecs/office_standards/ms-oe376/1663dabc-5d98-463f-889e-bcd9b77c3d34. Rather, the .fntdata files seem to contain rasterized glyphs. There is a decent codebase for doing the rendering, too (see, for example, https://github.com/foliojs/fontkit), *HOWEVER* that obviously contextual in terms of how the font is used in the actual .pptx. Since we don't know how MS's proprietary code goes about things, it'll be very difficult to replicate. 

It's a shame, would have been too easy...

**Comment by gitbrent** [comment]
Thanks for the well-researched explanation. TIL :-)

**Comment by pimlottc-gov** [comment]
I wonder if a possible strategy could be to merge some generated slides with another pptx that already has the embedded font in it.  Although merging pptx files is its own challenge.

**Comment by jiangjiji** [comment]
https://github.com/kiwiwings/poi-font-mbender

I have found a useful project to embedd fonts.

**Comment by mathbruyen** [comment]
After doing some investigation we have been able to embed fonts in Powerpoint. Leaving here if that helps someone.

`.fntdata` files that Powerpoint uses is actually an alias for `.eot`, I couldn't confirm whether it supports the full format but by looking at the binary content of `.fntdata` files created by Powerpoint it looks the header follows https://www.w3.org/submissions/2008/SUBM-EOT-20080305/ (flags ang magic bytes found). So a ttf => eot converter should be able to produce expected files. Alternatively extracting such files from a generated Powerpoint file works fine (this is what we use).

Then there is a bit of work, that we reverse engineered from generating a file, opening it in Powerpoint, enabling font embedding (full character set), and save. We were able to script:

* enable font embedding by adding `embedTrueTypeFonts` attribute, remove `saveSubsetFonts` attribute (partial character set) in `ppt/presentation.xml`
* declare the font in `docProps/app.xml` (increase count + add name in list)
* for every font/variant to be embedded:
    * add content type for `.fntdata` in `[Content_Types].xml`
    * add a relation for the font in `ppt/_rels/presentation.xml.rels` (beware that relationship ids must be sequential)
    * dump `.fntdata` at ``fonts/font${nextFontFileId}.fntdata` (beware that ids must be sequential)
    * add embedded font to the `embeddedFontLst` list in `ppt/presentation.xml`
